src/lib-compression/istream-zstd.c

```diff
@@ -0,0 +1,237 @@
+/* Copyright (c) 2013-2018 Dovecot authors, see the included COPYING file */
+
+#include "lib.h"
+
+#ifdef HAVE_ZSTD
+
+#include "buffer.h"
+#include "istream-private.h"
+#include "istream-zlib.h"
+#include <zstd.h>
+
+struct zstd_istream {
+	struct istream_private istream;
+	ZSTD_DStream *dstream;
+
+	struct stat last_parent_statbuf;
+	ZSTD_inBuffer input;
+	size_t input_true_size;
+	ZSTD_outBuffer output;
+	size_t next_read;
+	ssize_t buffer_size;
+	bool marked : 1;
+	bool log_errors : 1;
+};
+
+static void i_stream_zstd_close(struct iostream_private *stream,
+				bool close_parent)
+{
+	struct zstd_istream *zstream = (struct zstd_istream *)stream;
+
+	ZSTD_freeDStream(zstream->dstream);
+	zstream->dstream = NULL;
+
+	i_free(zstream->input.src);
```

Comment:
This produces a warning in gcc 8.2.1 

```
istream-zstd.c:34:23: warning: passing argument 2 of ‘p_free_internal’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  i_free(zstream->input.src);
```
Any suggestions on how to resolve it without making a temporary pointer?